draft-ietf-jsonpath-base.md

```diff
@@ -1207,19 +1200,161 @@ compared is an object, array, boolean, or `null`.
 * The comparison `a > b` yields true if and only if `b < a` yields true.
 * The comparison `a >= b` yields true if and only if `b < a` yields true or `a == b` yields true.
 
-##### Regular Expressions
+##### Boolean Operators
 {: unnumbered}
 
-A regular-expression test yields true if and only if the value on the left-hand side of `=~` is a string value and it
-matches the regular expression on the right-hand side according to the semantics of {{-iregexp}}.
+The logical AND, OR, and NOT operators have the normal semantics of Boolean algebra and
+obey its laws (see, for example, {{BOOLEAN-LAWS}}).
 
-The semantics of regular expressions are as defined in {{-iregexp}}.
+##### Function Extensions {#fnex}
+{: unnumbered}
 
-##### Boolean Operators
+[^unnumbered-bad]
+
+[^unnumbered-bad]: This should not be an unnumbered section, which it is
+    forced to be by the parent section being unnumbered too.
+    The overall structure where the parts of this subsection go to
+    needs to be decided; one part already is in the IANA
+    Considerations in {{iana-fnex}}.
+    This PR will focus on content, a future PR will fix the structure.
+{:source=" -- cabo"}
+
+Beyond the filter expression functionality defined in the preceding
+subsections, JSONPath defines an extension point that can be used to
+add filter expression functionality: "Function Extensions".
+
+A function extension defines a registered name (see {{iana-fnex}}) that
+can be applied to a sequence of zero or more arguments, resulting in a
+result.
+A function-expression stands for ("returns") a nodelist or a value.
+The arguments are filter expressions that are to be interpreted as
+nodelists or values.
+For each argument and for the result, the function extension defines
+the "kind" of the item, i.e., whether the item is a value or a
+nodelist ("nodes").
+
+~~~ abnf
+function-name           = function-name-first *function-name-char
+function-name-first     = LCALPHA
+function-name-char      = DIGIT / function-name-first / "_"
+LCALPHA                 = %x61-7A  ; "a".."z"
+
+
+function-expression     = function-name "(" S [function-argument
+                             *(S "," S function-argument)] S ")"
+singular-path           =/ function-expression ; if value
+filter-path             =/ function-expression ; if nodelist
+function-argument       = comparable / filter-path
+~~~
+
+Syntactically, a function-expression can occur anywhere where a
+singular-path can occur (exist-expr, comparable).
+A function-expression that employs a function extension that returns a
+nodelist MUST return a singular node to be used in a comparable.
```

Comment:
Why would it not be acceptable to return an empty nodelist since that's valid for use in a comparable?

Reply:
Separately: what should the behaviour be if a nodelist with multiple nodes is returned in this situation?

Reply:
One way to handle this would be to build a type system with values, singular paths (which are implicitly converted into values where a value is required), and filter-paths.
Functions already have to define the types of their arguments and their result, so this could be checked statically.

Reply:
Yes, I like that. The more static checks we can get, the fewer runtime errors the spec will need to cope with.